packages/hardhat-core/src/builtin-tasks/compile.ts

```diff
@@ -1016,78 +1005,186 @@ subtask(TASK_COMPILE_SOLIDITY_GET_COMPILATION_JOBS_FAILURE_REASONS)
+            `File ${sourceName} doesn't match any of the configured compilers`
+          );
+
+          errorMessage += `* ${sourceName} (${versionsRange})\n`;
+        }
+
+        errorMessage += "\n";
       }
+
       if (importsIncompatibleFile.length > 0) {
-        reasons += `These files import other files that use a different and incompatible version of Solidity:
+        errorMessage += `These files import other files that use a different and incompatible version of Solidity:
+
+`;
+
+        for (const error of importsIncompatibleFile) {
+          const { sourceName } = error.file;
+          const { versionPragmas } = error.file.content;
+          const versionsRange = versionPragmas.join(" ");
+
+          const incompatibleImportsFiles: ResolvedFile[] =
+            error.extra?.incompatibleImports ?? [];
+
+          const incompatibleImports = incompatibleImportsFiles.map(
+            (x: ResolvedFile) =>
+              `${x.sourceName} (${x.content.versionPragmas.join(" ")})`
+          );
+
+          log(
+            `File ${sourceName} imports files ${incompatibleImportsFiles
+              .map((x) => x.sourceName)
+              .join(", ")} that use an incompatible version of Solidity`
+          );
+
+          let importsText = "";
+          if (incompatibleImports.length === 1) {
+            importsText = ` imports ${incompatibleImports[0]}`;
+          } else if (incompatibleImports.length === 2) {
+            importsText = ` imports ${incompatibleImports[0]} and ${incompatibleImports[1]}`;
+          } else if (incompatibleImports.length > 2) {
+            const otherImportsCount = incompatibleImports.length - 2;
+            importsText = ` imports ${incompatibleImports[0]}, ${
+              incompatibleImports[1]
+            } and ${otherImportsCount} other ${pluralize(
+              otherImportsCount,
+              "file"
+            )}. Use --verbose to see the full list.`;
+          }
+
+          errorMessage += `* ${sourceName} (${versionsRange})${importsText}\n`;
+        }
 
-${importsIncompatibleFile.map((x) => `* ${x}`).join("\n")}
+        errorMessage += "\n";
+      }
+
+      if (indirectlyImportsIncompatibleFile.length > 0) {
+        errorMessage += `These files depend on other files that use a different and incompatible version of Solidity:
 
 `;
+
+        for (const error of indirectlyImportsIncompatibleFile) {
+          const { sourceName } = error.file;
+          const { versionPragmas } = error.file.content;
+          const versionsRange = versionPragmas.join(" ");
+
+          const incompatibleImportsFiles: ResolvedFile[] =
+            error.extra?.incompatibleIndirectImports ?? [];
+
+          const incompatibleImports = incompatibleImportsFiles.map(
+            (x: ResolvedFile) =>
+              `${x.sourceName} (${x.content.versionPragmas.join(" ")})`
+          );
+
+          log(
+            `File ${sourceName} depends on files ${incompatibleImportsFiles
```

Comment:
We could add the entire import path here, can't we?

Something like a second line with "Import path: A.sol > B.sol > Incompatible.sol"